fuse_optimizers/src/fixed_lag_smoother.cpp

```diff
@@ -255,12 +256,58 @@ void FixedLagSmoother::processQueue(fuse_core::Transaction& transaction)
-  if (!pending_transactions_.empty())
+  const auto current_time = pending_transactions_.front().stamp();
+
+  // If we just started because an ignition sensor transaction was received, we try to process it individually. This is
+  // important because we need to update the graph with the ignition sensor transaction in order to get the motion
+  // models notified of the initial state. The motion models will typically maintain a state history in order to create
+  // motion model constraints with the optimized variables from the updated graph. If we do not process the ignition
+  // sensor transaction individually, the motion model constraints created for the other queued transactions will not be
+  // able to use any optimized variables from the graph because it is not been optimized yet, and they will have to use
+  // a default zero state instead. This can easily lead to local minima because the variables in the graph are not
+  // initialized properly, i.e. they do not take the ignition sensor transaction into account.
+  if (ignited_)
   {
-    current_time = pending_transactions_.front().stamp();
+    // The ignition sensor transaction is assumed to be at the end of the queue, because it must be the oldest one.
+    // If there is more than one ignition sensor transaction in the queue, it is always the oldest one that started
+    // things up.
+    ignited_ = false;
+
+    const auto transaction_rbegin = pending_transactions_.rbegin();
+    const auto& element = *transaction_rbegin;
+    if (!std::binary_search(params_.ignition_sensors.begin(), params_.ignition_sensors.end(), element.sensor_name))
+    {
+      // We just started, but the oldest transaction is not from an ignition sensor. We will still process the
+      // transaction, but we do not enforce it is processed individually.
+      ROS_ERROR_STREAM("The queued transaction with timestamp " << element.stamp() << " from sensor " <<
+                       element.sensor_name << " is not an ignition sensor transaction. " <<
+                       "This transaction will not be processed individually.");
+    }
+    else if (applyMotionModels(element.sensor_name, *element.transaction))
+    {
+      // Processing was successful. Add the results to the final transaction, delete this one, and return, so the
+      // transaction from the ignition sensor is processed individually.
+      transaction.merge(*element.transaction, true);
+      erase(pending_transactions_, transaction_rbegin);
+      return;
+    }
+    else
+    {
+      // The motion model processing failed. When this happens to an ignition sensor transaction there is no point on
+      // trying again next time, so we ignore this transaction.
+      ROS_ERROR_STREAM("The queued ignition transaction with timestamp " << element.stamp() << " from sensor " <<
+                       element.sensor_name << " could not be processed. Ignoring this transaction.");
+      erase(pending_transactions_, transaction_rbegin);
```

Comment:
Since the ignition transaction was never applied, I feel like we should also set `started_` to false. That way we don't start using transactions on an improperly initialized system.

There is a nasty edge-case in here as well, when the pending transactions include a second ignition transaction. To be very thorough we would need to crawl through any remaining transactions and check if it is an ignition source.
* If so, then purge the transactions before it and leave with `started_ = true` and `ignited_ = true`. This would trigger an ignition attempt during the next optimization cycle.
* If not, then leave the queue as-is, and exit with `started_ = false` and `ignited_ = false`. The next received ignition transaction will then trigger an ignition attempt.

Reply:
First of all, when I wrote this I obviously thought of the edge cases, but I didn't test most of them. :smiley: 

> Since the ignition transaction was never applied, I feel like we should also set `started_` to false. That way we don't start using transactions on an improperly initialized system.

The reason I didn't do that is because `processQueue` is called from the optimization loop, and it looks to me it's too late to stop processing things.

Here it looks like we should do:
``` c++
started_ = false;
pending_transactions_.clear();
```

With this we'll end up with an empty `new_transaction` in the optimization loop: https://github.com/locusrobotics/fuse/blob/devel/fuse_optimizers/src/fixed_lag_smoother.cpp#L198

Maybe there we could make `processQueue` return a `bool` to indicate if there's something to process, or we could just check if `new_transaction` has any constraints or is empty. We could also check the `started_` flag. It looks like any of those options would work. My point here is that I think we also need to add some changes to the optimization loop, to abort processing the queue.

> There is a nasty edge-case in here as well, when the pending transactions include a second ignition transaction. To be very thorough we would need to crawl through any remaining transactions and check if it is an ignition source.

I thought about this, and my conclusion was that a second ignition transaction wouldn't produce an ignition cycle. The first one would set `started_ = true`, so the second one wouldn't be able to enter https://github.com/locusrobotics/fuse/blob/devel/fuse_optimizers/src/fixed_lag_smoother.cpp#L367. I was thinking of the case where an ignition sensor model is used both to generate ignition and non-ignition transactions.

But I think you're right. I the first ignition fails, then we need to purge all transactions started from the first ignition till the next ignition, if any. If there's a second one, then we can either process it in this cycle, or leave for the next one. I believe it's possible to process it in this cycle, but it's likely cleaner to simply purge things in this cycle and process in the next one, so we only have to do this (instead of the code snippet above in this comment):

``` c++
// Purge all transaction after the ignition one that just failed. But if we find another ignition
// transaction, we schedule it to be processed in the next optimization cycle
const auto pending_ignition_transaction_iter =
    std::find_if(pending_transactions_.rbegin(), pending_transactions_.rend(),
        [&params_](const auto& element)
        {
          return std::binary_search(params_.ignition_sensors.begin(),
                                                   params_.ignition_sensors.end(),
                                                   element.sensor_name);
        });

if (pending_ignition_transaction_iter == pending_transactions_.rend())
{
   pending_transactions_.clear();
   started_ = false;
}
else
{
  if (pending_transaction_iter != pending_transactions_.rbegin())
  {
    erase(pending_transactions_, std::prev(pending_ignition_transaction_iter));
  }
  ignited_ = true;
  return;
}
```

And, as I mentioned before, we probably need to do changes in the optimization loop, so it properly handles "empty" transactions, if it currently doesn't.

Is this more or less what you're thinking about?

I can update the PR with this changes, but since you already have a unit test, I guess you can do that yourself, or I wait for you to push the unit test to this or another PR. Thanks for adding a unit test, btw! :smiley:

Reply:
Yeah, that's the kind of thing I was thinking about.

I'm not sure you need to clear all the pending transactions in the first `if`. When a new ignition transaction is received, the pending transactions will get cleared as needed. I would think removing the failed ignition transaction would be sufficient.

And then handling the empty transaction case, which I can do in a separate PR if you can add the above block to this PR.

Reply:
> I'm not sure you need to clear all the pending transactions in the first if. When a new ignition transaction is received, the pending transactions will get cleared as needed. I would think removing the failed ignition transaction would be sufficient.

Sure, instead of clearing the pending transactions, I could just `return;`

Just saw your message. Give me a minute to update this PR. Thanks.

Reply:
Updated.

However, bear in mind I haven't tested the edges cases this logic handles.